modules/ROOT/pages/extending-neo4j/procedures.adoc

```diff
@@ -190,9 +190,94 @@ The classes that can be injected are:
 * `Transaction`
 //* `SecurityContext`
 //* `ProcedureTransaction`
+//* `ProcedureMemory` Candidate for public API but not stable yet
 
 All of the above classes are considered safe and future-proof and do not compromise the security of the database.
 Several unsupported (restricted) classes can also be injected and can be changed with little or no notice.
 Procedures written to use these restricted APIs are not loaded by default, and you need to use the `dbms.security.procedures.unrestricted` to load unsafe procedures.
 Read more about this config setting in link:{neo4j-docs-base-uri}/operations-manual/{page-version}/security/securing-extensions[Operations Manual -> Securing extensions].
 
+[[memory-resource-tracking]]
+== Memory Resource Tracking
+
+[NOTE]
+====
+The memory resource tracking API for procedure framework is available for preview.
+Future versions of neo4j might contain breaking changes to this API.
+====
+
+If your procedure allocates significant amounts of heap memory,
+you can register allocations to count towards the
+link:{neo4j-docs-base-uri}/operations-manual/{page-version}/performance/memory-configuration/#memory-configuration-limit-transaction-memory[configured transaction limits].
+This allows you to avoid `OutOfMemory` errors that cause database restarts.
+Memory allocations will also show up in query profiles.
+
+To do this you need to inject a `org.neo4j.procedure.memory.ProcedureMemory` as a field in your procedure class.
+`ProcedureMemory` has various methods to allow you to register allocations.
+For example (see javadocs for a full reference):
```

Comment:
Unfortunately it's not possible to link to the java docs until this has become public API, but I'll try to remember to insert a link in here then.